---
title: "OS: scRNAseq from external set"
author: "by Alex"
output: 
  html_document:
  toc: TRUE
toc_float: FALSE
editor_options: 
  chunk_output_type: console
---
  
  
  <style type="text/css">
  .main-container {
    max-width: 1800px;
    margin-left: 20px;
    margin-right: auto;
  }
.toc-content {
  max-width: 1800px;
  margin-left: 50px;
  margin-right: auto;
}

div {
  
  margin-left: 20px;
}


hr.new1 {
  border-top: 1px solid #84a8e0;
}



</style>
  
  # Results {.tabset}
  
  
  <div style="background-color: #f0f5c1;">
  
  
  </div>
  
  <hr>
  
  
  
```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE, include = FALSE , cache.lazy = FALSE )

library ( dplyr)
library(scales)
library("RColorBrewer")
library("ggplot2")
library(gplots)
library(ggrepel)
library(Hmisc)
library(stringr)
library(gridExtra)
library(knitr)
library(kableExtra)
library ( openxlsx)
library ( ggbeeswarm )
library(Seurat)
library ( HGNChelper )

lapply(c("ggraph","igraph","tidyverse", "data.tree"), library, character.only = T)



options(scipen=999)
getPalette = colorRampPalette(brewer.pal(9, "Set1")) # expand color pallete
source ( "accessory.R")

```


```{r}

# after preprocessing and merging will be stored here
preprocessed_dir = "/projects/lab.mis/ASC_projects/scRNAseq_zhou_etal/pre_processed/"
dir.create ( preprocessed_dir )
filename = paste0(preprocessed_dir, "zhou_OS.rds" )

normalizeded_dir = "/projects/lab.mis/ASC_projects/scRNAseq_zhou_etal/normalized/"
dir.create ( normalizeded_dir )
filename_norm = paste0(normalizeded_dir, "zhou_OS.rds" )




# there are 5 samples 
raw_dir = "/projects/lab.mis/ASC_projects/scRNAseq_zhou_etal/raw/"
samples = list.dirs(raw_dir)
samples = samples[ grepl ( "BC", samples)]

raw_df = data.frame()
for ( s in samples ){
  #id = str_match(s, ".*_(.*)\\.matrix")[, 2]
  id = str_match(s, ".*\\/(.*)$")[, 2]
  raw_df = rbind ( raw_df , data.frame ( pid=id, dir=s))
}




```



```{r}
mt_cut = 15 
nfeature_min = 200 
nfeature_max = 4000


```

# Summary {.tabset}

  * The goal of this is to prepare single cell OS samples from zhou et al, 
    + https://www.nature.com/articles/s41467-020-20059-6
  * There are 11 samples that we downloaded from the GEO ftp site: accession GSE152048
  * We run Prepocessing and QC to filter out low quality cells. 
  * alls samples were merged for downstream analysis 

## pre-processing workflow {.tabset}

  * The steps below encompass the standard pre-processing workflow based on Seurat's Basic recommendations. However some paremeters might be tweaked for the specific experiments. 
  
  * Cells are filtered based on 2 main factors.
  
    + cells that have unique feature counts over `r nfeature_max` or less than `r nfeature_min`
    
      - Low-quality cells or empty droplets will often have very few genes
      -  Cell doublets or multiplets may exhibit an aberrantly high gene count
      - __Thus we try to find a Goldilock range: not too low but not too high__
    
    + cells with > `r mt_cut `% mitochondrial counts. High mitochondrial indicates Low-quality / dying cells.
    
      -Low-quality / dying cells often exhibit extensive mitochondrial contamination
      - We calculate mitochondrial QC metrics with the `PercentageFeatureSet` function, which calculates the percentage of counts originating from a set of features
      - We use the set of all genes starting with `MT-` as a set of mitochondrial genes

```{r}

# preprocess each of the samples 
# each sample will be store din a list 

slist = list () # each sample will be store in this list to be merged later. 
preplots = list()
raw_df$before_preprocess = 1
raw_df$after_preprocess = 1



for ( r in 1:nrow ( raw_df) ){

# r=1 
pid = raw_df[r:r, ]
id = pid$pid
# Load the data
data <- Read10X(data.dir = pid$dir )
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = data, project = pid$pid, min.cells = 3, min.features = 200)
# add mitochondria percent 
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
head(pbmc@meta.data, 5)

# plot mitochondria 
mt_plot <- plot_hist_perc ( pbmc@meta.data, plot_this="percent.mt", ccc="#6e9be5" , cut_here = 15, msg = "" , binwidth=5, y=.3) 
nfeature_plot <- plot_hist_perc ( pbmc@meta.data, plot_this="nFeature_RNA", ccc="#6e9be5" 
                                  , cut_here = c (nfeature_min,  nfeature_max)
                                  , msg = "" 
                                  , binwidth=200, y=.02  # y is the position of annotated text
                                  ) 

preplots[[id]][["hist"]] <- (  ( mt_plot + ggtitle(pid$pid) ) / nfeature_plot )

# quantile ( pbmc@meta.data$percent.mt) # here we see that 50% of the cells will be removed if MT is set to 10% 
# quantile ( pbmc@meta.data$nFeature_RNA) # 25% will be removed if the high end is set to 2500 



# Visualize QC metrics as a violin plot
before_plot = VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, cols="#FAF632" )
before_total = length ( Cells(pbmc) )

# after 
pbmc =  subset(pbmc, subset = nFeature_RNA > nfeature_min & nFeature_RNA < nfeature_max & percent.mt < mt_cut )
after_total = length ( Cells(pbmc) )
after_plot = VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, cols="#526E9F")


preplots[[id]][["befor_after"]]  =  (before_plot  ) / after_plot 

raw_df[r, ]$before_preprocess = before_total
raw_df[r, ]$after_preprocess = after_total

# save seurat object
slist[[id]] = pbmc

print ( paste ( "completed", id ))  
}


```

























