---
title: "OS: scRNAseq from external set"
author: "by Alex"
output: 
  html_document:
  toc: TRUE
toc_float: FALSE
editor_options: 
  chunk_output_type: console
---
  
  
  <style type="text/css">
  .main-container {
    max-width: 1800px;
    margin-left: 20px;
    margin-right: auto;
  }
.toc-content {
  max-width: 1800px;
  margin-left: 50px;
  margin-right: auto;
}

div {
  
  margin-left: 20px;
}


hr.new1 {
  border-top: 1px solid #84a8e0;
}



</style>
  
  # Results {.tabset}
  
  
  <div style="background-color: #f0f5c1;">
  
  
  </div>
  
  <hr>
  
  
  
```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE, include = FALSE , cache.lazy = FALSE )

library ( dplyr)
library(scales)
library("RColorBrewer")
library("ggplot2")
library(gplots)
library(ggrepel)
library(Hmisc)
library(stringr)
library(gridExtra)
library(knitr)
library(kableExtra)
library ( openxlsx)
library ( ggbeeswarm )
library(Seurat)
library ( HGNChelper )
library(harmony)
lapply(c("ggraph","igraph","tidyverse", "data.tree"), library, character.only = T)



options(scipen=999)
getPalette = colorRampPalette(brewer.pal(9, "Set1")) # expand color pallete
source ( "accessory.R")

```


```{r}

# after preprocessing and merging will be stored here
preprocessed_dir = "/projects/lab.mis/ASC_projects/scRNAseq_zhou_etal/pre_processed/"
dir.create ( preprocessed_dir )
filename = paste0(preprocessed_dir, "zhou_OS.rds" )

normalizeded_dir = "/projects/lab.mis/ASC_projects/scRNAseq_zhou_etal/normalized/"
dir.create ( normalizeded_dir )
filename_norm = paste0(normalizeded_dir, "zhou_OS.rds" )




# there are 5 samples 
raw_dir = "/projects/lab.mis/ASC_projects/scRNAseq_zhou_etal/raw/"
samples = list.dirs(raw_dir)
samples = samples[ grepl ( "BC", samples)]

raw_df = data.frame()
for ( s in samples ){
  #id = str_match(s, ".*_(.*)\\.matrix")[, 2]
  id = str_match(s, ".*\\/(.*)$")[, 2]
  raw_df = rbind ( raw_df , data.frame ( pid=id, dir=s))
}




```



```{r}
mt_cut = 15 
nfeature_min = 200 
nfeature_max = 4000


```

# Summary {.tabset}

  * The goal of this is to prepare single cell OS samples from zhou et al, 
    + https://www.nature.com/articles/s41467-020-20059-6
  * There are __11 samples__ total that we downloaded from the GEO ftp site: accession GSE152048
  * We run Prepocessing and QC to filter out low quality cells. 
  * alls samples were merged and pre-processed together. 
  * Finally Harmony was used to integrate all samples with each individual sample as a batch. 

## pre-processing workflow {.tabset}

  * The steps below encompass the standard pre-processing workflow based on Seurat's Basic recommendations. However some paremeters might be tweaked for the specific experiments. 
  
  * Cells are filtered based on 2 main factors.
  
    + cells that have unique feature counts over `r nfeature_max` or less than `r nfeature_min`
    
      - Low-quality cells or empty droplets will often have very few genes
      -  Cell doublets or multiplets may exhibit an aberrantly high gene count
      - __Thus we try to find a Goldilock range: not too low but not too high__
    
    + cells with > `r mt_cut `% mitochondrial counts. High mitochondrial indicates Low-quality / dying cells.
    
      -Low-quality / dying cells often exhibit extensive mitochondrial contamination
      - We calculate mitochondrial QC metrics with the `PercentageFeatureSet` function, which calculates the percentage of counts originating from a set of features
      - We use the set of all genes starting with `MT-` as a set of mitochondrial genes

```{r}

# preprocess each of the samples 
# each sample will be store din a list 

slist = list () # each sample will be store in this list to be merged later. 
preplots = list()
raw_df$before_preprocess = 1
raw_df$after_preprocess = 1



for ( r in 1:nrow ( raw_df) ){

# r=1 
pid = raw_df[r:r, ]
id = pid$pid
# Load the data
data <- Read10X(data.dir = pid$dir )
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = data, project = pid$pid, min.cells = 3, min.features = 200)
# add mitochondria percent 
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
head(pbmc@meta.data, 5)

# plot mitochondria 
mt_plot <- plot_hist_perc ( pbmc@meta.data, plot_this="percent.mt", ccc="#6e9be5" , cut_here = 15, msg = "" , binwidth=5, y=.3) 
nfeature_plot <- plot_hist_perc ( pbmc@meta.data, plot_this="nFeature_RNA", ccc="#6e9be5" 
                                  , cut_here = c (nfeature_min,  nfeature_max)
                                  , msg = "" 
                                  , binwidth=200, y=.02  # y is the position of annotated text
                                  ) 

preplots[[id]][["hist"]] <- (  ( mt_plot + ggtitle(pid$pid) ) / nfeature_plot )

# quantile ( pbmc@meta.data$percent.mt) # here we see that 50% of the cells will be removed if MT is set to 10% 
# quantile ( pbmc@meta.data$nFeature_RNA) # 25% will be removed if the high end is set to 2500 



# Visualize QC metrics as a violin plot
before_plot = VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, cols="#FAF632" )
before_total = length ( Cells(pbmc) )

# after 
pbmc =  subset(pbmc, subset = nFeature_RNA > nfeature_min & nFeature_RNA < nfeature_max & percent.mt < mt_cut )
after_total = length ( Cells(pbmc) )
after_plot = VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, cols="#526E9F")


preplots[[id]][["befor_after"]]  =  (before_plot  ) / after_plot 

raw_df[r, ]$before_preprocess = before_total
raw_df[r, ]$after_preprocess = after_total

# save seurat object
slist[[id]] = pbmc

print ( paste ( "completed", id ))  
}


```




### tabulate. 

  * before and after shows how many cells were kept after preprocessing. 


```{r fig=TRUE,fig.width=16, fig.height=10, echo=FALSE, include=TRUE, results='asis' }
temp = raw_df 
temp$dir = NULL 

kable( temp, format = "html" , row.names = F, caption = "Samples" ) %>% kable_classic(full_width = F, position = "center")

```


```{r fig=TRUE,fig.width=17, fig.height=10, echo=FALSE, include=TRUE, results='asis' }

for ( n in names ( preplots)){
  cat ( "### ", n, "\n\n"  )
  print ( ( preplots[[n]]$hist)  |  (preplots[[n]]$befor_after ))
  cat ( '\n\n')
}

```

```{r}


# final merging 

names ( slist )


if ( ! file.exists ( filename_norm) ){




pbmc.big <-  merge(slist[["BC10" ]]
                  , y = c(slist[["BC11"]], slist[["BC16"]]
                          , slist[[ "BC17" ]], slist[[ "BC2" ]]
                          , slist[[ "BC20" ]], slist[[ "BC21" ]]
                          , slist[[ "BC22" ]], slist[[ "BC3" ]]
                          , slist[[ "BC5" ]] , slist[[ "BC6" ]]
                          )
                  
                  , add.cell.ids = names ( preplots)
                 , project = "zhou_OS")


saveRDS(pbmc.big, filename  ) # we save this before normalizing because we may need normalize again when combining with JMML samples 

unique ( pbmc.big@meta.data$orig.ident)

# this are the data 
head  ( pbmc.big@assays$RNA@data[ 1:3, 1:3] )
head ( pbmc.big@meta.data)
cell =  pbmc.big@meta.data$orig.ident


pbmc.big[["exp"]] <- cell
# here we need to also correct for batch and mitochondria. Cell cycles 
# no need to correct for exp since we are going to use harmony to do it later
#pbmc.big <- SCTransform(pbmc.big, vst.flavor = "v2", verbose = TRUE, vars.to.regress=  c("percent.mt","exp")  ) 
pbmc.big <- SCTransform(pbmc.big, vst.flavor = "v2", verbose = TRUE, vars.to.regress=  c("percent.mt"  )  ) 
# save before clustering... does'nt matter but it shoul save some space. 

pbmc.big <- pbmc.big %>% 
    RunPCA(npcs = 30, verbose = FALSE) %>%
    RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
    FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
    FindClusters(resolution = 0.7, verbose = FALSE)
# this is how many clusters this thing yield 


pbmc_harmony <- pbmc.big %>% RunHarmony("exp", plot_convergence = T)



 pbmc_harmony <-    pbmc_harmony %>% 
    # ScaleData(features = rownames(pbmc)) %>%, not needed since SCTransform should had done this already 
    # Scaled data is here pbmc.big[["SCT"]]@scale.data[1:3, 1:3 ]
    RunPCA(  npcs = 30, verbose = FALSE) %>%
    RunUMAP(reduction = "harmony", dims = 1:30, verbose = FALSE) %>%
    FindNeighbors(reduction = "harmony", dims = 1:30, verbose = FALSE) %>%
    FindClusters(resolution = 0.7, verbose = FALSE)


saveRDS( pbmc_harmony, filename_norm ) 
} else {
  pbmc_harmony = readRDS(filename_norm )
}



```


```{r}


# this is how many clusters this thing yield 
clusters = levels ( pbmc_harmony )
ccc_clusters = getPalette( length ( clusters) )


DimPlot(pbmc_harmony, label = T, repel = T ) # plots total - default for seurat transformation is 1:30 so no need to mess with this. 
# changing resolution does'nt seem to effect this experiment

# want to visualize if there is a batch 

batch_plot = DimPlot(pbmc_harmony, label = T, repel = T, group.by = "orig.ident") + 
  ggtitle("Unsupervised clustering") +  scale_color_manual(values = getPalette( 50) )



```



```{r}

# auto assign 
ccc = readRDS(
  "/ehome/resource/random_tools/colors_21.rds"
)
source("/ehome/resource/annotation/single_cell/ScType/gene_sets_prepare.R")
source("/ehome/resource/annotation/single_cell/ScType/sctype_score_.R")
db_ = "/ehome/resource/annotation/single_cell/ScType/addendum_bone.xlsx";

tissue = c ( "bone") # e.g. Immune system, Liver, Pancreas, Kidney, Eye, Brain

# prepare gene sets
gs_list = gene_sets_prepare(db_, tissue)

es.max = sctype_score(scRNAseqData = pbmc_harmony[["SCT"]]@scale.data, scaled = TRUE, 
                      gs = gs_list$gs_positive, gs2 = gs_list$gs_negative) 

# merge by cluster

cL_resutls = do.call("rbind", lapply(unique(pbmc_harmony@meta.data$seurat_clusters), function(cl){
  es.max.cl = sort(rowSums(es.max[ ,rownames(pbmc_harmony@meta.data[pbmc_harmony@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
  head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(pbmc_harmony@meta.data$seurat_clusters==cl)), 10)
}))
sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  


# set low-confident (low ScType score) clusters to "unknown"
sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = "Unknown"
print(sctype_scores[,1:3])


pbmc_harmony@meta.data$customclassif = ""
for(j in unique(sctype_scores$cluster)){
  cl_type = sctype_scores[sctype_scores$cluster==j,]; 
  pbmc_harmony@meta.data$customclassif[pbmc_harmony@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])
}

#ccc = setNames( getPalette( length ( unique ( unique ( pbmc_harmony@meta.data$customclassif)) ) )
 #                              , unique ( pbmc_harmony@meta.data$customclassif)
  #                             )

unique  ( pbmc_harmony@meta.data[pbmc_harmony@meta.data$customclassif == "osteoclast",  ]$seurat_clusters )
unique  ( pbmc_harmony@meta.data[pbmc_harmony@meta.data$customclassif == "myeloid",  ]$seurat_clusters )


pbmc_harmony@meta.data$celltype = ifelse ( pbmc_harmony@meta.data$seurat_clusters == 13, "osteoclast", "_" )
pbmc_harmony@meta.data$celltype = ifelse ( pbmc_harmony@meta.data$seurat_clusters == 6, "myeloid",  pbmc_harmony@meta.data$celltype  )


cell_type_plot = DimPlot(pbmc_harmony, reduction = "umap", label = FALSE, repel = F, group.by = 'customclassif', seed=2
                          )    +  
  scale_color_manual(values = ccc )

cell_type_plot = LabelClusters(cell_type_plot, id = "customclassif", size = 5, repel = T) + ggtitle ( " Predicted Cell type")

```



## Visuals {.tabset}

* merged the samples into a single assay and normalized/scaled and removed unwanted variation with the `SCTransform` function. 
* We corrected for Mitochondria.
* final integration was completed with harmony. 


### Overall After scaling and normalization 

  * sample names are colored to indicate any batch effect even after integration

```{r fig=TRUE,fig.width=10, fig.height=8, echo=FALSE, include=TRUE, results='asis' }
batch_plot
```

### Cell type

  * we used an automated cell-type indentifier ScType to label and group each clusters. 
  * Ianevski et al, 2022
  * https://www.nature.com/articles/s41467-022-28803-w
  * for cell type defintion we merged annotation markers from the zhou et al paper + CellMarker Zhang et al
    + Zhang et al we selected bone as the sample source.  
    + after merging the zhou and zhang annotations any other cell type that were missing, eg Endothelial cell we extract from the CellMarker by selecting the most recurrent gene for a specific cell type, irresptive of sample source. 


```{r fig=TRUE,fig.width=10, fig.height=8, echo=FALSE, include=TRUE, results='asis' }
cell_type_plot + ggtitle ( "cell type")
```




















